File: backend/recipe_diversity.py
"""Detect and enforce genuinely different recipes on regeneration."""

from __future__ import annotations

import re

COOKING_METHOD_RULES: list[tuple[str, tuple[str, ...]]] = [
    ("baked", (r"קרם אפוי", r"\bbaked\b", r"בתנור", r"אופ", r"\boven\b")),
    ("chilled", (r"מוס", r"\bmousse\b", r"קר(?:ה|ים)", r"ללא אפייה", r"\bno[- ]?bake\b", r"\bchill")),
    ("pancake", (r"פנקייק", r"חבית", r"\bpancake")),
    ("pudding", (r"פודינג", r"\bpudding")),
    ("fried", (r"מטוג", r"\bfry", r"\bfried", r"במחבת")),
    ("boiled", (r"מרתיח", r"מבשל", r"\bboil", r"\bsimmer")),
    ("cream", (r"קרם(?! אפוי)", r"\bcream\b")),
]

DESSERT_CATEGORY_RULES: list[tuple[str, tuple[str, ...]]] = [
    ("baked_custard", (r"קרם אפוי", r"flan", r"crème brûlée")),
    ("mousse", (r"מוס", r"\bmousse")),
    ("pancake", (r"פנקייק", r"חבית", r"\bpancake")),
    ("pudding", (r"פודינג", r"\bpudding")),
    ("cake", (r"עוג", r"\bcake")),
    ("cookies", (r"עוגיות", r"ביסקוויט", r"\bcookie")),
    ("cream", (r"קרם", r"\bcream")),
    ("brownie", (r"בראוניז", r"\bbrownie")),
    ("ice_cream", (r"גלידה", r"\bice cream")),
]


def _normalize_title(title: str) -> str:
    return re.sub(r"\s+", " ", (title or "").strip().lower())


def _match_rule(text: str, rules: list[tuple[str, tuple[str, ...]]]) -> str | None:
    for rule_id, patterns in rules:
        if any(re.search(pattern, text, flags=re.I) for pattern in patterns):
            return rule_id
    return None


def detect_cooking_method(recipe: dict) -> str:
    text = f"{recipe.get('name', '')}\n" + "\n".join(recipe.get("steps") or [])
    return _match_rule(text, COOKING_METHOD_RULES) or "general"


def detect_dessert_category(recipe: dict) -> str:
    text = f"{recipe.get('name', '')}\n" + "\n".join(recipe.get("steps") or [])
    return _match_rule(text, DESSERT_CATEGORY_RULES) or "general"


def is_duplicate_title(title: str, exclude_titles: list[str]) -> bool:
    normalized = _normalize_title(title)
    return any(_normalize_title(item) == normalized for item in exclude_titles)


def validate_recipe_diversity(
    recipe: dict,
    *,
    recipe_type: str = "meal",
    exclude_titles: list[str] | None = None,
    exclude_cooking_methods: list[str] | None = None,
    exclude_dessert_categories: list[str] | None = None,
) -> dict:
    exclude_titles = exclude_titles or []
    exclude_cooking_methods = exclude_cooking_methods or []
    exclude_dessert_categories = exclude_dessert_categories or []
    failures: list[str] = []
    cooking_method = detect_cooking_method(recipe)
    dessert_category = detect_dessert_category(recipe) if recipe_type == "dessert" else None

    if exclude_titles and is_duplicate_title(recipe.get("name", ""), exclude_titles):
        failures.append("duplicate_title")
    if (
        exclude_cooking_methods
        and cooking_method in exclude_cooking_methods
        and cooking_method != "general"
    ):
        failures.append("duplicate_cooking_method")
    if (
        recipe_type == "dessert"
        and exclude_dessert_categories
        and dessert_category in exclude_dessert_categories
        and dessert_category != "general"
    ):
        failures.append("duplicate_dessert_category")

    return {
        "ok": not failures,
        "failures": failures,
        "cooking_method": cooking_method,
        "dessert_category": dessert_category,
    }


def build_regeneration_prompt_section(
    *,
    language: str = "he",
    exclude_titles: list[str] | None = None,
    exclude_cooking_methods: list[str] | None = None,
    exclude_dessert_categories: list[str] | None = None,
) -> str:
    exclude_titles = exclude_titles or []
    exclude_cooking_methods = exclude_cooking_methods or []
    exclude_dessert_categories = exclude_dessert_categories or []
    if not exclude_titles and not exclude_cooking_methods and not exclude_dessert_categories:
        return ""

    titles = ", ".join(item for item in exclude_titles if item)
    methods = ", ".join(item for item in exclude_cooking_methods if item)
    categories = ", ".join(item for item in exclude_dessert_categories if item)

    if language == "he":
        return f"""
כללי יצירת מתכון חדש (חובה — לא לחזור על מתכון קודם):
- צור מנה שונה לגמרי — לא אותו מתכון עם שם אחר.
- אל תחזור על שמות: {titles or '(אין)'}.
- אל תחזור על שיטת הכנה: {methods or '(אין)'}.
- אל תחזור על סוג קינוח: {categories or '(אין)'}.
- השתמש באותם מרכיבים זמינים בלבד, אך בנה מנה אחרת (למשל: אם היה "קרם וניל אפוי" — אפשר "מוס וניל קר", "פנקייק וניל" או "פודינג וניל", אך לא אותה מנה).
- החזר מתכון אחד חדש בלבד.
"""

    return f"""
NEW RECIPE RULES (mandatory — do not repeat a previous dish):
- Create a genuinely different dish — not the same recipe with a new title.
- Do NOT repeat titles: {titles or '(none)'}.
- Do NOT repeat cooking methods: {methods or '(none)'}.
- Do NOT repeat dessert categories: {categories or '(none)'}.
- Use the same available ingredients only, but build a different dish.
- Return one new recipe only.
"""


def _main_ingredient_label(ingredients: list[str], language: str) -> str:
    first = (ingredients or [""])[0]
    bare = re.sub(
        r"^[\d\s/]+(?:כפ(?:ית|ות)|כ(?:ף|פות)|גרם|מ\"ל|כוס(?:ות)?|יח(?:ידה|ידות)?)\.?\s*",
        "",
        str(first),
        flags=re.I,
    ).strip()
    if not bare:
        return "מתוק" if language == "he" else "sweet"
    return re.sub(r"\(.*?\)", "", bare).strip()


DESSERT_STYLE_VARIANTS: list[dict] = [
    {
        "id": "mousse",
        "category": "mousse",
        "method": "chilled",
        "title_he": lambda main: f"מוס {main}",
        "title_en": lambda main: f"{main} mousse",
        "steps_he": lambda main, mins: [
            f"מכניסים את {main} לקערה ומקציפים בעזרת מערבל ידני עד לתערובת קלילה.",
            "מוסיפים את שאר המרכיבים ומערבבים בעדינות עד לקבלת מרקם אחיד.",
            f"מעבירים לקערות הגשה ומקררים במקרר כ-{mins} דקות עד שהמוס מתייצב.",
            "מגישים קר.",
        ],
        "steps_en": lambda main, mins: [
            f"Whip {main} in a bowl until light and airy.",
            "Fold in the remaining ingredients gently until smooth.",
            f"Chill in serving cups for about {mins} minutes until set.",
            "Serve cold.",
        ],
    },
    {
        "id": "pancake",
        "category": "pancake",
        "method": "fried",
        "title_he": lambda main: f"פנקייק {main}",
        "title_en": lambda main: f"{main} pancakes",
        "steps_he": lambda main, mins: [
            f"מערבבים את {main} עם שאר המרכיבים בקערה עד לבלילה חלקה.",
            "מחממים מחבת על אש בינונית ומשמנים קלות.",
            f"מוזגים כף גדולה לכל פנקייק ומטגנים כ-{max(2, round(mins / 4))} דקות מכל צד עד להזהבה.",
            "מגישים חם.",
        ],
        "steps_en": lambda main, mins: [
            f"Mix {main} with the remaining ingredients until smooth.",
            "Warm a lightly oiled pan over medium heat.",
            f"Cook each pancake for about {max(2, round(mins / 4))} minutes per side until golden.",
            "Serve warm.",
        ],
    },
    {
        "id": "pudding",
        "category": "pudding",
        "method": "boiled",
        "title_he": lambda main: f"פודינג {main}",
        "title_en": lambda main: f"{main} pudding",
        "steps_he": lambda main, mins: [
            f"מערבבים את {main} עם שאר המרכיבים בסיר על אש נמוכה.",
            f"מבשלים תוך ערבוב רציף כ-{mins} דקות עד שהתערובת מסמיכה.",
            "מעבירים לקערות הגשה ומצננים מעט לפני ההגשה.",
            "מגישים בטמפרטורת החדר או מעט פושר.",
        ],
        "steps_en": lambda main, mins: [
            f"Combine {main} with the remaining ingredients in a saucepan over low heat.",
            f"Cook, stirring constantly, for about {mins} minutes until thickened.",
            "Pour into serving bowls and cool slightly before serving.",
            "Serve at room temperature or slightly warm.",
        ],
    },
    {
        "id": "baked_custard",
        "category": "baked_custard",
        "method": "baked",
        "title_he": lambda main: f"קרם {main} אפוי",
        "title_en": lambda main: f"Baked {main} custard",
        "steps_he": lambda main, mins: [
            f"מערבבים את {main} עם שאר המרכיבים עד לתערובת חלקה.",
            "יוצקים לתבניות קטנות ומניחים בתוך תבנית עם מים (אמבט מים).",
            f"אופים בתנור ב-170°C כ-{mins} דקות עד שהקרם מוצק אך רך.",
            "מצננים מעט ומגישים.",
        ],
        "steps_en": lambda main, mins: [
            f"Whisk {main} with the remaining ingredients until smooth.",
            "Pour into ramekins and place in a water bath.",
            f"Bake at 340°F for about {mins} minutes until set but tender.",
            "Cool slightly and serve.",
        ],
    },
    {
        "id": "cream",
        "category": "cream",
        "method": "chilled",
        "title_he": lambda main: f"קרם {main}",
        "title_en": lambda main: f"{main} cream",
        "steps_he": lambda main, mins: [
            f"מערבבים את {main} עם שאר המרכיבים עד לקרם אחיד.",
            "מעבירים לקערת הגשה ומיישרים את פני השטח.",
            f"מקררים כ-{mins} דקות עד שהקרם מתייצב.",
            "מגישים קר.",
        ],
        "steps_en": lambda main, mins: [
            f"Mix {main} with the remaining ingredients until creamy and smooth.",
            "Transfer to a serving dish and level the top.",
            f"Refrigerate for about {mins} minutes until set.",
            "Serve chilled.",
        ],
    },
]


def pick_alternate_dessert_variant(
    *,
    ingredients: list[str] | None = None,
    language: str = "he",
    cooking_time: int = 30,
    exclude_titles: list[str] | None = None,
    exclude_cooking_methods: list[str] | None = None,
    exclude_dessert_categories: list[str] | None = None,
) -> dict:
    from dessert_dish_title import build_dessert_dish_title, pick_primary_flavor_label

    exclude_titles = exclude_titles or []
    exclude_cooking_methods = exclude_cooking_methods or []
    exclude_dessert_categories = exclude_dessert_categories or []
    main = pick_primary_flavor_label(ingredients or [], language)
    mins = min(cooking_time, max(15, round(cooking_time * 0.6)))

    available = []
    for variant in DESSERT_STYLE_VARIANTS:
        built = build_dessert_dish_title(
            ingredients or [],
            language=language,
            style_id=variant["id"],
            exclude_titles=exclude_titles,
            exclude_cooking_methods=exclude_cooking_methods,
            exclude_dessert_categories=exclude_dessert_categories,
        )
        if is_duplicate_title(built["name"], exclude_titles):
            continue
        if variant["method"] in exclude_cooking_methods:
            continue
        if variant["category"] in exclude_dessert_categories:
            continue
        available.append(variant)

    variant = available[0] if available else DESSERT_STYLE_VARIANTS[0]
    built = build_dessert_dish_title(
        ingredients or [],
        language=language,
        style_id=variant["id"],
        exclude_titles=exclude_titles,
        exclude_cooking_methods=exclude_cooking_methods,
        exclude_dessert_categories=exclude_dessert_categories,
    )
    steps = (
        variant["steps_he"](main, mins) if language == "he" else variant["steps_en"](main, mins)
    )
    return {
        "name": built["name"],
        "steps": steps,
        "cooking_method": variant["method"],
        "dessert_category": variant["category"],
        "style_id": variant["id"],
    }

File: src/utils/recipeDiversity.js
/**
 * Detect and enforce genuinely different recipes on regeneration.
 */

import {
  buildDessertDishTitle,
  isDuplicateDessertTitle,
  pickPrimaryFlavorLabel,
} from './dessertDishTitle'
import {
  formatEnglishStepIngredientList,
  formatHebrewStepIngredientList,
  toStepIngredientReference,
} from './recipeStepWording'
import { canonicalIngredient } from '../data/ingredientKnowledge'

function normalizeTitle(title) {
  return String(title ?? '')
    .trim()
    .toLowerCase()
    .replace(/\s+/g, ' ')
}

const COOKING_METHOD_RULES = [
  { id: 'baked', patterns: [/קרם אפוי/i, /\bbaked\b/i, /בתנור/i, /אופ(?:ים|ה|י(?:ה|ים))/i, /\boven\b/i] },
  { id: 'chilled', patterns: [/מוס/i, /\bmousse\b/i, /קר(?:ה|ים)/i, /ללא אפייה/i, /\bno[- ]?bake\b/i, /\bchill/i] },
  { id: 'pancake', patterns: [/פנקייק/i, /חבית(?:י)?(?:ת|ות)?/i, /\bpancake/i] },
  { id: 'pudding', patterns: [/פודינג/i, /\bpudding/i] },
  { id: 'fried', patterns: [/מטוג(?:ן|נ)/i, /\bfry|\bfried/i, /במחבת/i] },
  { id: 'boiled', patterns: [/מרתיח/i, /מבשל/i, /\bboil|\bsimmer/i] },
  { id: 'cream', patterns: [/קרם(?! אפוי)/i, /\bcream\b/i] },
]

const DESSERT_CATEGORY_RULES = [
  { id: 'baked_custard', patterns: [/קרם אפוי/i, /flan/i, /crème brûlée/i] },
  { id: 'mousse', patterns: [/מוס/i, /\bmousse/i] },
  { id: 'pancake', patterns: [/פנקייק/i, /חבית/i, /\bpancake/i] },
  { id: 'pudding', patterns: [/פודינג/i, /\bpudding/i] },
  { id: 'cake', patterns: [/עוג(?:ה|יות)/i, /\bcake/i] },
  { id: 'cookies', patterns: [/עוגיות/i, /ביסקוויט/i, /\bcookie/i] },
  { id: 'cream', patterns: [/קרם/i, /\bcream/i] },
  { id: 'brownie', patterns: [/בראוניז/i, /\bbrownie/i] },
  { id: 'ice_cream', patterns: [/גלידה/i, /\bice cream/i] },
]

function matchRule(text, rules) {
  const haystack = String(text ?? '')
  for (const rule of rules) {
    if (rule.patterns.some((pattern) => pattern.test(haystack))) {
      return rule.id
    }
  }
  return null
}

export function detectCookingMethod(recipe) {
  const title = recipe?.name ?? ''
  const stepsText = (recipe?.steps ?? []).join('\n')
  return matchRule(`${title}\n${stepsText}`, COOKING_METHOD_RULES) ?? 'general'
}

export function detectDessertCategory(recipe) {
  const title = recipe?.name ?? ''
  const stepsText = (recipe?.steps ?? []).join('\n')
  return matchRule(`${title}\n${stepsText}`, DESSERT_CATEGORY_RULES) ?? 'general'
}

export function isDuplicateTitle(title, excludeTitles = []) {
  const normalized = normalizeTitle(title)
  return excludeTitles.some((item) => normalizeTitle(item) === normalized)
}

export function validateRecipeDiversity(recipe, {
  recipeType = 'meal',
  excludeTitles = [],
  excludeCookingMethods = [],
  excludeDessertCategories = [],
} = {}) {
  const failures = []
  const cookingMethod = detectCookingMethod(recipe)
  const dessertCategory = recipeType === 'dessert' ? detectDessertCategory(recipe) : null

  if (excludeTitles.length && isDuplicateTitle(recipe?.name, excludeTitles)) {
    failures.push('duplicate_title')
  }
  if (
    excludeCookingMethods.length &&
    excludeCookingMethods.includes(cookingMethod) &&
    cookingMethod !== 'general'
  ) {
    failures.push('duplicate_cooking_method')
  }
  if (
    recipeType === 'dessert' &&
    excludeDessertCategories.length &&
    dessertCategory &&
    excludeDessertCategories.includes(dessertCategory) &&
    dessertCategory !== 'general'
  ) {
    failures.push('duplicate_dessert_category')
  }

  return {
    ok: failures.length === 0,
    failures,
    cookingMethod,
    dessertCategory,
  }
}

const DESSERT_STEP_VARIANTS = [
  {
    id: 'mousse',
    category: 'mousse',
    method: 'chilled',
    stepsHe: (main, mins) => [
      `מכניסים את ${main} לקערה ומקציפים בעזרת מערבל ידני עד לתערובת קלילה.`,
      'מוסיפים את שאר המרכיבים ומערבבים בעדינות עד לקבלת מרקם אחיד.',
      `מעבירים לקערות הגשה ומקררים במקרר כ-${mins} דקות עד שהמוס מתייצב.`,
      'מגישים קר.',
    ],
    stepsEn: (main, mins) => [
      `Whip ${main} in a bowl until light and airy.`,
      'Fold in the remaining ingredients gently until smooth.',
      `Chill in serving cups for about ${mins} minutes until set.`,
      'Serve cold.',
    ],
  },
  {
    id: 'pancake',
    category: 'pancake',
    method: 'fried',
    stepsHe: (main, mins) => [
      `מערבבים את ${main} עם שאר המרכיבים בקערה עד לבלילה חלקה.`,
      'מחממים מחבת על אש בינונית ומשמנים קלות.',
      `מוזגים כף גדולה לכל פנקייק ומטגנים כ-${Math.max(2, Math.round(mins / 4))} דקות מכל צד עד להזהבה.`,
      'מגישים חם.',
    ],
    stepsEn: (main, mins) => [
      `Mix ${main} with the remaining ingredients until smooth.`,
      'Warm a lightly oiled pan over medium heat.',
      `Cook each pancake for about ${Math.max(2, Math.round(mins / 4))} minutes per side until golden.`,
      'Serve warm.',
    ],
  },
  {
    id: 'pudding',
    category: 'pudding',
    method: 'boiled',
    stepsHe: (main, mins) => [
      `מערבבים את ${main} עם שאר המרכיבים בסיר על אש נמוכה.`,
      `מבשלים תוך ערבוב רציף כ-${mins} דקות עד שהתערובת מסמיכה.`,
      'מעבירים לקערות הגשה ומצננים מעט לפני ההגשה.',
      'מגישים בטמפרטורת החדר או מעט פושר.',
    ],
    stepsEn: (main, mins) => [
      `Combine ${main} with the remaining ingredients in a saucepan over low heat.`,
      `Cook, stirring constantly, for about ${mins} minutes until thickened.`,
      'Pour into serving bowls and cool slightly before serving.',
      'Serve at room temperature or slightly warm.',
    ],
  },
  {
    id: 'baked_custard',
    category: 'baked_custard',
    method: 'baked',
    stepsHe: (main, mins) => [
      `מערבבים את ${main} עם שאר המרכיבים עד לתערובת חלקה.`,
      'יוצקים לתבניות קטנות ומניחים בתוך תבנית עם מים (אמבט מים).',
      `אופים בתנור ב-170°C כ-${mins} דקות עד שהקרם מוצק אך רך.`,
      'מצננים מעט ומגישים.',
    ],
    stepsEn: (main, mins) => [
      `Whisk ${main} with the remaining ingredients until smooth.`,
      'Pour into ramekins and place in a water bath.',
      `Bake at 340°F for about ${mins} minutes until set but tender.`,
      'Cool slightly and serve.',
    ],
  },
  {
    id: 'cream',
    category: 'cream',
    method: 'chilled',
    stepsHe: (main, mins) => [
      `מערבבים את ${main} עם שאר המרכיבים עד לקרם אחיד.`,
      'מעבירים לקערת הגשה ומיישרים את הפני השטח.',
      `מקררים כ-${mins} דקות עד שהקרם מתייצב.`,
      'מגישים קר.',
    ],
    stepsEn: (main, mins) => [
      `Mix ${main} with the remaining ingredients until creamy and smooth.`,
      'Transfer to a serving dish and level the top.',
      `Refrigerate for about ${mins} minutes until set.`,
      'Serve chilled.',
    ],
  },
  {
    id: 'cups',
    category: 'cream',
    method: 'chilled',
    stepsHe: (main, mins) => [
      `מערבבים את ${main} עם שאר המרכיבים עד לקרם אחיד.`,
      'ממלאים כוסות הגשה קטנות בשכבות או בקרם אחיד.',
      `מקררים כ-${mins} דקות עד שהקינוח מתייצב.`,
      'מגישים קר.',
    ],
    stepsEn: (main, mins) => [
      `Mix ${main} with the remaining ingredients until smooth.`,
      'Fill small serving cups with the mixture.',
      `Chill for about ${mins} minutes until set.`,
      'Serve cold.',
    ],
  },
]

export function pickAlternateDessertVariant({
  ingredients = [],
  language = 'he',
  cookingTime = 30,
  excludeTitles = [],
  excludeCookingMethods = [],
  excludeDessertCategories = [],
} = {}) {
  const mins = Math.min(cookingTime, Math.max(15, Math.round(cookingTime * 0.6)))
  const main = pickPrimaryFlavorLabel(ingredients, language)

  const available = DESSERT_STEP_VARIANTS.filter((variant) => {
    const built = buildDessertDishTitle(ingredients, {
      language,
      styleId: variant.id,
      excludeTitles,
      excludeCookingMethods,
      excludeDessertCategories,
    })
    if (isDuplicateDessertTitle(built.name, excludeTitles)) return false
    if (excludeCookingMethods.includes(variant.method)) return false
    if (excludeDessertCategories.includes(variant.category)) return false
    return true
  })

  const variant = available[0] ?? DESSERT_STEP_VARIANTS[0]
  const built = buildDessertDishTitle(ingredients, {
    language,
    styleId: variant.id,
    excludeTitles,
    excludeCookingMethods,
    excludeDessertCategories,
  })
  const steps = language === 'he' ? variant.stepsHe(main, mins) : variant.stepsEn(main, mins)

  return {
    name: built.name,
    steps,
    cookingMethod: variant.method,
    dessertCategory: variant.category,
    styleId: variant.id,
  }
}

function buildIngredientListPhrase(ingredients, language = 'he') {
  const names = (ingredients ?? []).map((item) => String(item ?? '').trim()).filter(Boolean)
  const refs = names.map((name) => toStepIngredientReference(name, language))
  return language === 'he'
    ? formatHebrewStepIngredientList(refs)
    : formatEnglishStepIngredientList(refs)
}

function buildGenericMealSteps(listPhrase, cookingTime, language, style = 'skillet') {
  const mins = Math.min(cookingTime, Math.max(10, Math.round(cookingTime / 2)))
  if (style === 'pot') {
    return language === 'he'
      ? [
          'מחממים סיר על אש בינונית.',
          `מוסיפים את ${listPhrase} ומבשלים יחד תוך ערבוב מדי פעם.`,
          `ממשיכים כ-${mins} דקות עד שהמרכיבים רכים ומשתלבים.`,
          'טועמים, מתבלים לפי הצורך ומגישים חם.',
        ]
      : [
          'Warm a pot over medium heat.',
          `Add ${listPhrase} and cook together, stirring occasionally.`,
          `Continue for about ${mins} minutes until tender and combined.`,
          'Taste, adjust seasoning, and serve hot.',
        ]
  }
  return language === 'he'
    ? [
        'מחממים מחבת על אש בינונית עם מעט שמן.',
        `מוסיפים את ${listPhrase} ומבשלים תוך ערבוב.`,
        `ממשיכים כ-${mins} דקות עד שהמרכיבים משתלבים ומקבלים צבע.`,
        'מגישים חם.',
      ]
    : [
        'Heat a lightly oiled pan over medium heat.',
        `Add ${listPhrase} and cook, stirring often.`,
        `Continue for about ${mins} minutes until combined and fragrant.`,
        'Serve hot.',
      ]
}

function buildPastaEggSteps(listPhrase, cookingTime, language, style = 'homestyle') {
  const pastaMins = Math.min(cookingTime, Math.max(8, Math.round(cookingTime * 0.35)))
  const finishMins = Math.min(cookingTime, Math.max(4, Math.round(cookingTime * 0.2)))
  if (style === 'skillet') {
    return language === 'he'
      ? [
          `מבשלים את הפסטה במים רותחים עם מלח כ-${pastaMins} דקות עד al dente, מסננים ושומרים בצד.`,
          'מחממים מחבת על אש בינונית עם מעט שמן.',
          `מוסיפים את ${listPhrase} למחבת ומבשלים יחד כ-${finishMins} דקות.`,
          'מגישים מיד.',
        ]
      : [
          `Boil the pasta in salted water for about ${pastaMins} minutes until al dente, then drain.`,
          'Warm a pan with a little oil over medium heat.',
          `Add ${listPhrase} to the pan and cook together for about ${finishMins} minutes.`,
          'Serve right away.',
        ]
  }
  if (style === 'rustic') {
    return language === 'he'
      ? [
          `מבשלים את הפסטה במים רותחים עם מלח כ-${pastaMins} דקות עד al dente, מסננים ושומרים מעט ממי הבישול.`,
          `מחזירים את הפסטה לסיר, מוסיפים את ${listPhrase} ומערבבים בעדינות.`,
          `מחממים על אש נמוכה כ-${finishMins} דקות עד שהמנה אחידה.`,
          'מגישים חם.',
        ]
      : [
          `Cook the pasta in salted boiling water for about ${pastaMins} minutes until al dente; drain, reserving a little water.`,
          `Return the pasta to the pot, add ${listPhrase}, and toss gently.`,
          `Warm over low heat for about ${finishMins} minutes until combined.`,
          'Serve hot.',
        ]
  }
  return language === 'he'
    ? [
        `מבשלים את הפסטה במים רותחים עם מלח כ-${pastaMins} דקות עד al dente, מסננים ושומרים בצד.`,
        `מחממים מחבת, מוסיפים את ${listPhrase} ומערבבים.`,
        `מחזירים את הפסטה למחבת ומבשלים יחד כ-${finishMins} דקות.`,
        'מגישים חם.',
      ]
    : [
        `Boil the pasta in salted water for about ${pastaMins} minutes until al dente, then drain.`,
        `Warm a pan, add ${listPhrase}, and stir.`,
        `Return the pasta to the pan and cook together for about ${finishMins} minutes.`,
        'Serve hot.',
      ]
}

export function pickAlternateMealVariant({
  ingredients = [],
  language = 'he',
  cookingTime = 30,
  excludeTitles = [],
  excludeCookingMethods = [],
} = {}) {
  const listPhrase = buildIngredientListPhrase(ingredients, language)
  const set = new Set(
    (ingredients ?? []).map((item) => canonicalIngredient(String(item))).filter(Boolean),
  )
  const hasPasta = set.has('pasta')
  const hasEgg = set.has('egg') || set.has('eggs')
  const variants =
    hasPasta && hasEgg
      ? [
          {
            id: 'pasta-egg-homestyle',
            method: 'general',
            name: language === 'he' ? 'פסטה קרמית בסגנון ביתי' : 'Creamy Homestyle Pasta',
            steps: buildPastaEggSteps(listPhrase, cookingTime, language, 'homestyle'),
          },
          {
            id: 'pasta-egg-skillet',
            method: 'fried',
            name: language === 'he' ? 'פסטה מהירה בשמן זית' : 'Quick Olive Oil Pasta',
            steps: buildPastaEggSteps(listPhrase, cookingTime, language, 'skillet'),
          },
          {
            id: 'pasta-egg-rustic',
            method: 'boiled',
            name: language === 'he' ? 'פסטה עם ביצה מקושקשת' : 'Pasta with Scrambled Egg',
            steps: buildPastaEggSteps(listPhrase, cookingTime, language, 'rustic'),
          },
        ]
      : [
          {
            id: 'meal-skillet',
            method: 'fried',
            name: language === 'he' ? 'מוקפץ ירק מהיר' : 'Quick Vegetable Sauté',
            steps: buildGenericMealSteps(listPhrase, cookingTime, language, 'skillet'),
          },
          {
            id: 'meal-pot',
            method: 'boiled',
            name: language === 'he' ? 'תבשיל ביתי בסיר' : 'Homestyle Pot Stew',
            steps: buildGenericMealSteps(listPhrase, cookingTime, language, 'pot'),
          },
          {
            id: 'meal-pan',
            method: 'general',
            name: language === 'he' ? 'מנה חמה מהמטבח' : 'Warm Kitchen-Style Dish',
            steps: buildGenericMealSteps(listPhrase, cookingTime, language, 'skillet'),
          },
        ]
  const available = variants.filter(
    (variant) =>
      !isDuplicateTitle(variant.name, excludeTitles) &&
      !excludeCookingMethods.includes(variant.method),
  )
  return (
    available[0] ??
    variants.find((variant) => !isDuplicateTitle(variant.name, excludeTitles)) ??
    variants[0]
  )
}

export function buildRegenerationPromptSection({
  language = 'he',
  excludeTitles = [],
  excludeCookingMethods = [],
  excludeDessertCategories = [],
} = {}) {
  if (!excludeTitles.length && !excludeCookingMethods.length && !excludeDessertCategories.length) {
    return ''
  }

  const titles = excludeTitles.filter(Boolean).join(', ')
  const methods = excludeCookingMethods.filter(Boolean).join(', ')
  const categories = excludeDessertCategories.filter(Boolean).join(', ')

  if (language === 'he') {
    return `
כללי יצירת מתכון חדש (חובה — לא לחזור על מתכון קודם):
- צור מנה שונה לגמרי — לא אותו מתכון עם שם אחר.
- אל תחזור על שמות: ${titles || '(אין)'}.
- אל תחזור על שיטת הכנה: ${methods || '(אין)'}.
- אל תחזור על סוג קינוח: ${categories || '(אין)'}.
- השתמש באותם מרכיבים זמינים בלבד, אך בנה מנה אחרת (למשל: אם היה "קרם וניל אפוי" — אפשר "מוס וניל קר", "פנקייק וניל" או "פודינג וניל", אך לא אותה מנה).
- החזר מתכון אחד חדש בלבד.
`
  }

  return `
NEW RECIPE RULES (mandatory — do not repeat a previous dish):
- Create a genuinely different dish — not the same recipe with a new title.
- Do NOT repeat titles: ${titles || '(none)'}.
- Do NOT repeat cooking methods: ${methods || '(none)'}.
- Do NOT repeat dessert categories: ${categories || '(none)'}.
- Use the same available ingredients only, but build a different dish (e.g. if the first was baked vanilla custard, try vanilla mousse, vanilla pancakes, or vanilla pudding — never the same dish).
- Return one new recipe only.
`
}
